api/src/main/java/io/minio/credentials/IamAwsProvider.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * MinIO Java SDK for Amazon S3 Compatible Cloud Storage, (C) 2020 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.minio.credentials;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.MapperFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.minio.messages.ResponseDate;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Objects;
+import javax.annotation.Nullable;
+import okhttp3.HttpUrl;
+import okhttp3.OkHttpClient;
+import okhttp3.Protocol;
+import okhttp3.Request;
+import okhttp3.Response;
+
+/**
+ * Credential provider using <a
+ * href="http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html">IAM roles
+ * for Amazon EC2</a>.
+ */
+public class IamAwsProvider extends EnvironmentProvider {
+  // Custom endpoint to fetch IAM role credentials.
+  private final HttpUrl customEndpoint;
+  private final OkHttpClient httpClient;
+  private final ObjectMapper mapper;
+  private Credentials credentials;
+
+  public IamAwsProvider(@Nullable String customEndpoint, @Nullable OkHttpClient customHttpClient) {
+    this.customEndpoint =
+        (customEndpoint != null)
+            ? Objects.requireNonNull(HttpUrl.parse(customEndpoint), "Invalid custom endpoint")
+            : null;
+    // HTTP/1.1 is only supported in default client because of HTTP/2 in OkHttpClient cause 5
+    // minutes timeout on program exit.
+    this.httpClient =
+        (customHttpClient != null)
+            ? customHttpClient
+            : new OkHttpClient().newBuilder().protocols(Arrays.asList(Protocol.HTTP_1_1)).build();
+    this.mapper = new ObjectMapper();
+    this.mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+    this.mapper.configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true);
+  }
+
+  private void checkLoopbackHost(HttpUrl url) {
+    try {
+      for (InetAddress addr : InetAddress.getAllByName(url.host())) {
+        if (!addr.isLoopbackAddress()) {
+          throw new IllegalArgumentException(url.host() + " is not loopback only host");
```

Comment:
Not sure if `IllegalArgumentException` is the right choice for all the exceptions in this class.

Reply:
Sorry I wrote this comment at the wrong place with the wrong exception name! I meant to say that I'm not sure if `IllegalStateException` makes sense for most of the exceptions in the class..